### DISCUSSION TITLE
Provide an imperative API for injection head content

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#541] [Proposal]
### Body

# Summary

Provide an imperative API for inserting head content (styles, scripts, etc) for use-cases where they might not be a `<head>` element.

# Background & Motivation

With the introduction of content collections which lazily adds styles it has become more difficult to maintain implicit head injection. Today in CC you are required to have a `<head>` element in the page that uses it.

In 3.0 it would be good to completely remove implicit head injection. An imperative API for cases where a `<head>` doesn't exist would be a good replacement.

# Goals

- Allow manually getting head content so that it can be inserted as the developer pleases.
- Allow putting head content some where other than the bottom of the head; perhaps you might want it before some other scripts you have.
- Allow head content to be generated for API routes.
- Allow smaller demo pages where you only need to have a single island and don't want to create all of the repetitive wrapper html/head tags.

# Example

Something like:

```astro
---
import Component from '../components/Component.astro';
---

{Astro.injectHead()}

<Component client:load />
```

Bikeshed:

- `Astro.injectHead()`
- `Astro.headContent()`
- `Astro.head()`

**Comment by lilnasy** [comment]
... or `<Astro.Head/>`.

**Comment by lilnasy** [comment]
Relevant open issues:
- https://github.com/withastro/astro/issues/7641
- https://github.com/withastro/astro/issues/7610

Maybe we could have a "strict" mode, which would be partial by default - `<Astro.Head/>` and `<!DOCTYPE html>` would need to be added manually.
